### DISCUSSION TITLE
node-matter verses matternode?

### DISCUSSION THREAD

**Discussion by Hedda** [mfucci/node-matter#134] [Q&A] [answered]
Does https://github.com/mfucci/node-matter have different goals and use cases than https://github.com/project-chip/matternode or?

Not sure if FAQ from node-matter indirectly answers that or not(?):

https://github.com/mfucci/node-matter/blob/main/README.md#faq

> * _**Why using node-matter instead of the official codebase?**_
>   * _Well, the original codebase is platform dependent, has finicky tool version requirements and is over 8GB with all dependencies. This tool is less than 500kB and works on anything supporting node. Sure, it supports only the barebone Matter protocol for now._

**Comment by mfucci** [comment, marked as answer]
Hi Hedda,

Good question!

[project-chip/matternode](https://github.com/project-chip/matternode) is a recent addition in the official codebase and I haven't had the chance to update the Faq yet.

I discussed with the author of matternode (Martin Turon) and we agree that we have goals and use cases similar enough that made join our forces.
The result is [matter.js](https://github.com/project-chip/matter.js), a pure ECMSA script implementation of Matter (not using node.js APIs) that we will use as the core for node-matter and matternode and that will probably absorb most of the content of both projects.

node-matter might survive as the node.js layer on top of matter.js. The current is that the migration to matter.js should not interrupt other efforts on node-matter or matternode. It might migrate to the official codebase.

At this point:

- node-matter offers a fully operational Matter client / server that can be used to build a custom Matter device or control other Matter devices. End goal is to be a library that can be used to make a real Matter device in a few line of codes.
- matternode offers an advanced command line interface to start multiple Matter devices (client only), mostly for automated testing.

Both are highly experimental... meaning not a single API is stable.

What is your use case?
I can recommend you what will be your best bet at this point? :-)

Thanks!
- Marco

**Reply by Hedda** [reply]
> What is your use case?
> I can recommend you what will be your best bet at this point? :-)

The primary use case that can I see is for Zigbee2MQTT users as hoping that its developer considers a general implementation of node-matter or matternode as a client-only Zigbee-to-Matter bridge, preferably as a native component of the Zigbee2MQTT project. 

One of the main goals I would like to see achieved someday is for less advanced users of Zigbee2MQTT to have a much easier integrated option to install Zigbee2MQTT as an appliance and activate an embedded Matter client functionality inside it as a user-friendly interface that is part of the Zigbee2MQTT application (or at least part of the front-end from a users point-of-view), with the Matter protocol as an alternative point of integration instead of using MQTT, i.e. completely removing the need to use MQTT and an external MQTT broker in order to control the devices from other third-party ecosystems as long as they support Matter, e.g. such as example Apple iOS (Apple Home) and Google Home, (which today both currently require you to use some kind of third-party MQTT bridge application that act as the integration between Apple's and Google's smart home ecosystems)..

See these for reference:

 https://github.com/Koenkk/zigbee2mqtt/discussions/7443

https://www.zigbee2mqtt.io/guide/usage/integrations.html

FYI, Google also announced that the upcoming Google Home app update will bring an Script Editor feature for advanced routines, so users of Google Home ecosystem will be able to create and run automation directly in Google Home (via a Google Nest Hub Gen2) as long as they can get their devices from Zigbee2MQTT into Google Home that can be enough for many people:

https://9to5google.com/2022/10/04/google-home-web-app/

PS: I do not myself have the coding skill to code such as a solution for Zigbee2MQTT, so will just have to hope that someone else does.

PPS: Sure I myself can workaround this by using for example Home Assistant or Node-RED to allow us to control Zigbee2MQTT devices via Google Home App and Google Assistant, but I am not only thinking about myself but also of those people who are not advanced users and whose use case would be to only control Zigbee2MQTT devices directly via Google Home App and Google Assistant, without jumping through hoops to get an MQTT bridge up and running to the Google Home ecosystem,
